### PR TITLE
Limit heap growth on Android

### PR DESCRIPTION
**Pull request by dnfield**
Android side of https://github.com/flutter/flutter/issues/63631

Need to figure out how to test this, as all the surrounding code is untested 😢

### CODE REVIEW

File: shell/platform/android/io/flutter/embedding/engine/loader/FlutterLoader.java
Comment (jason-simmons) [inline shell/platform/android/io/flutter/embedding/engine/loader/FlutterLoader.java:251]: Can `oldGenHeapSizeMegaBytes` be provided as a shell command line flag in `shell/common/switches.h`?

File: shell/platform/android/io/flutter/embedding/engine/FlutterJNI.java
Comment (xster) [inline shell/platform/android/io/flutter/embedding/engine/FlutterJNI.java:97]: Making all the static private SG. But rather than making this class 2 tiered which is a bit more confusing, why not just let classes that want to mock FlutterJNI just inject a FlutterJNI mock directly so all methods can be mocked?

Comment (dnfield) [reply shell/platform/android/io/flutter/embedding/engine/FlutterJNI.java:97]: It limits the scope of what `FlutterLoader` gets, making it a little easier to quickly see exactly what it expects to use  and how these things are related.

File: shell/platform/android/io/flutter/embedding/engine/loader/FlutterLoader.java
Comment (dnfield) [reply shell/platform/android/io/flutter/embedding/engine/loader/FlutterLoader.java:251]: It could but then we lose some static type checking here. Is there some reason you have in mind to do it that way?

Comment (jason-simmons) [reply shell/platform/android/io/flutter/embedding/engine/loader/FlutterLoader.java:251]: It would be more consistent with how other engine settings are assigned and would avoid a proliferation of `nativeInit` arguments

Comment (dnfield) [reply shell/platform/android/io/flutter/embedding/engine/loader/FlutterLoader.java]: Done

File: shell/platform/android/io/flutter/embedding/engine/FlutterJNI.java
Comment (dnfield) [reply shell/platform/android/io/flutter/embedding/engine/FlutterJNI.java:97]: IOW, writing tests for this won't leave me wondering which of the many methods on FlutterJNI I'm supposed to mock.

Comment (xster) [reply shell/platform/android/io/flutter/embedding/engine/FlutterJNI.java:97]: Jason might have the historical background. I believe we made all the methods in the FlutterJNI class grouped in one class instead of being sliced and diced per consumer based on least privilege principle for a reason (probably to have a single place to look at the interop schema). If so, I strongly believe we should have a single pattern in the codebase to be consistent. Either keep everything in one class (and remove statics which I support) or break it all up and parcel views of the JNI to its consumer based on what the consumer needs. 

Otherwise, it becomes arbitrary why I could mock the FlutterLoader's JNI loader but on some JNI calls, it still crashes. 

I think the cleanest way for your intended change is still to make all the statics normal instance methods for the public APIs and inject the whole FlutterJNI in tests.

File: shell/platform/android/io/flutter/FlutterInjector.java
Comment (xster) [inline shell/platform/android/io/flutter/FlutterInjector.java:78]: FWIW, this is a breaking change. I still haven't merged the website PR for the last time this broke so I'll hold off until we resolve this PR. https://github.com/flutter/website/pull/4561

**Review by xster — CHANGES_REQUESTED** [review]

File: shell/platform/android/io/flutter/embedding/engine/FlutterJNI.java
Comment (dnfield) [reply shell/platform/android/io/flutter/embedding/engine/FlutterJNI.java:97]: The actual native methods are still on Flutter JNI, which still keeps it easier for us to write the JNI bindings. 

I'm not really clear on why we would want to keep all of the Java facing methods on the same class. Flutter JNI will stick around as this sort of God-object, and everyone you give it to will be able to call all sorts of native methods that may not make sense. 

I don't work on this code very much, so I'm not going to put up a huge fight about this. If, seeing what I'm saying, you still believe these methods should all be exposed on the Java side via the same one object, I can easily refactor this PR to do that - although if I were working on this more I would more strongly object :)

Comment (dnfield) [reply shell/platform/android/io/flutter/embedding/engine/FlutterJNI.java]: Talked offline. I'll move out this part of the change here, we can track what to do with FlutterJNI overall in https://github.com/flutter/flutter/issues/66131

I've also made the static methods public again, but marked them as deprecated with pointers to the instance methods.

Comment (xster) [reply shell/platform/android/io/flutter/embedding/engine/FlutterJNI.java]: Ping me when it's ready again

Comment (dnfield) [reply shell/platform/android/io/flutter/embedding/engine/FlutterJNI.java]: It should be ready now :) The test failure is an infra flake.

File: shell/platform/android/io/flutter/FlutterInjector.java
Comment (xster) [inline shell/platform/android/io/flutter/FlutterInjector.java:133]: 🤦‍♀️ thanks

File: shell/platform/android/io/flutter/embedding/engine/FlutterJNI.java
Comment (xster) [inline shell/platform/android/io/flutter/embedding/engine/FlutterJNI.java:101]: great. Thanks for adding docs

Comment (xster) [inline shell/platform/android/io/flutter/embedding/engine/FlutterJNI.java:101]: ultra-nit: `Loads the libflutter.so Flutter engine native library.`?

File: shell/platform/android/io/flutter/embedding/engine/loader/FlutterLoader.java
Comment (xster) [inline shell/platform/android/io/flutter/embedding/engine/loader/FlutterLoader.java:63]: Docs

File: shell/platform/android/io/flutter/embedding/engine/loader/FlutterLoader.java
Comment (xster) [inline shell/platform/android/io/flutter/embedding/engine/loader/FlutterLoader.java:67]: replace all `flutterJNILoader` with just `flutterJNI`. Since there's nothing else called jni loader here.

File: shell/platform/android/io/flutter/embedding/engine/loader/FlutterLoader.java
Comment (xster) [inline shell/platform/android/io/flutter/embedding/engine/loader/FlutterLoader.java:240]: I don't think we actually set largeHeap="true" in our AndroidManifest template right? Should we?

File: shell/platform/android/io/flutter/embedding/engine/loader/FlutterLoader.java
Comment (xster) [inline shell/platform/android/io/flutter/embedding/engine/loader/FlutterLoader.java:241]: Is this equality right? I assume your total memory size needs to accommodate everything right? Native memory, old gen, new gen, whatever else the VM mmaps. The old gen can't take up everything I imagine.

File: shell/platform/android/test/io/flutter/embedding/engine/PluginComponentTest.java
Comment (xster) [inline shell/platform/android/test/io/flutter/embedding/engine/PluginComponentTest.java:36]: Cool, lg. Can you push some changes to https://github.com/flutter/website/pull/4561 directly for the new pattern?

File: shell/platform/android/test/io/flutter/embedding/engine/loader/FlutterLoaderTest.java
Comment (xster) [inline shell/platform/android/test/io/flutter/embedding/engine/loader/FlutterLoaderTest.java:21]: import specific classes if possible.

File: shell/platform/android/test/io/flutter/embedding/engine/loader/FlutterLoaderTest.java
Comment (xster) [inline shell/platform/android/test/io/flutter/embedding/engine/loader/FlutterLoaderTest.java:41]: ditto for names here

File: shell/platform/android/test/io/flutter/embedding/engine/loader/FlutterLoaderTest.java
Comment (xster) [inline shell/platform/android/test/io/flutter/embedding/engine/loader/FlutterLoaderTest.java:51]: can this be a separate test? Since it seems unrelated to the test above.

**Review by xster — COMMENTED** [review]
Thanks for making the changes.

File: shell/platform/android/io/flutter/embedding/engine/FlutterJNI.java
Comment (dnfield) [reply shell/platform/android/io/flutter/embedding/engine/FlutterJNI.java:101]: libflutter.so C++ library good?

File: shell/platform/android/io/flutter/embedding/engine/loader/FlutterLoader.java
Comment (dnfield) [reply shell/platform/android/io/flutter/embedding/engine/loader/FlutterLoader.java:63]: Done.

File: shell/platform/android/io/flutter/embedding/engine/loader/FlutterLoader.java
Comment (dnfield) [reply shell/platform/android/io/flutter/embedding/engine/loader/FlutterLoader.java:67]: Done, also made this @NonNullable and made the default ctor new one up.

File: shell/platform/android/io/flutter/embedding/engine/loader/FlutterLoader.java
Comment (dnfield) [reply shell/platform/android/io/flutter/embedding/engine/loader/FlutterLoader.java:240]: That's for the Java heap. This property is unaffected by that setting - it's what the Java heap would be set to if largeHeap was set to true.

I don't think we need largeHeap set to true, since we really don't do much with the Java heap.

File: shell/platform/android/io/flutter/embedding/engine/loader/FlutterLoader.java
Comment (dnfield) [reply shell/platform/android/io/flutter/embedding/engine/loader/FlutterLoader.java:241]: This is what we have control over.

I'm probably garbling this up a little bit, but basically new space is moved to old space once it survives collection, and things are put directly into old space if they're over a certain size threshhold. So basically, the majority of heap space is reserved for old space. New space is kept smaller so that new shortlived garbage can be collected more aggressively.

This will cause Dart to run collections much more aggressively as we approach this limit. @rmacnak-google - will this be a hard limit for the VM, or will it be possible to go over it?

File: shell/platform/android/io/flutter/embedding/engine/FlutterJNI.java
Comment (xster) [reply shell/platform/android/io/flutter/embedding/engine/FlutterJNI.java:101]: SG

File: shell/platform/android/io/flutter/embedding/engine/loader/FlutterLoader.java
Comment (dnfield) [reply shell/platform/android/io/flutter/embedding/engine/loader/FlutterLoader.java:241]: We might want to bump this up a bit if it's a hard limit now that I think about it. Android doesn't want us to use more than this if we're good citizens, but this value might be a little lower than what's really "safe" to use in practice.

File: shell/platform/android/io/flutter/embedding/engine/loader/FlutterLoader.java
Comment (xster) [reply shell/platform/android/io/flutter/embedding/engine/loader/FlutterLoader.java:240]: ah got it. Thanks

File: shell/platform/android/test/io/flutter/embedding/engine/loader/FlutterLoaderTest.java
Comment (dnfield) [reply shell/platform/android/test/io/flutter/embedding/engine/loader/FlutterLoaderTest.java:21]: Done

File: shell/platform/android/test/io/flutter/embedding/engine/loader/FlutterLoaderTest.java
Comment (dnfield) [reply shell/platform/android/test/io/flutter/embedding/engine/loader/FlutterLoaderTest.java:41]: Done

File: shell/platform/android/io/flutter/embedding/engine/loader/FlutterLoader.java
Comment (xster) [reply shell/platform/android/io/flutter/embedding/engine/loader/FlutterLoader.java:241]: I don't know what this does behind the scene. I guess the desired behavior is probably like if we start trying to collect when it's like 60% of this value, then it's probably safe. If it collects at this value, we're probably going to have trouble.

File: shell/platform/android/test/io/flutter/embedding/engine/loader/FlutterLoaderTest.java
Comment (dnfield) [reply shell/platform/android/test/io/flutter/embedding/engine/loader/FlutterLoaderTest.java:51]: Done

File: shell/platform/android/io/flutter/embedding/engine/loader/FlutterLoader.java
Comment (xster) [inline shell/platform/android/io/flutter/embedding/engine/loader/FlutterLoader.java:65]: this is nicer

**Review by xster — APPROVED** [review]
LG, thanks

File: shell/platform/android/test/io/flutter/embedding/engine/PluginComponentTest.java
Comment (dnfield) [reply shell/platform/android/test/io/flutter/embedding/engine/PluginComponentTest.java:36]: https://github.com/flutter/website/pull/4561/commits/301b89ba5bdd0dc41ee79f84e957cd597ee00de9

File: shell/platform/android/io/flutter/embedding/engine/loader/FlutterLoader.java
Comment (dnfield) [reply shell/platform/android/io/flutter/embedding/engine/loader/FlutterLoader.java:245]: This acts like a limit the Dart VM tries to keep us from approaching asymptotically.

But this value is much lower than the limit we use today, which is smething like 32GB (and probably more ram than any Android device out there has). The highest I've seen this value is 512gb, on phones that have ~4gb of ram.

Comment (dnfield) [reply shell/platform/android/io/flutter/embedding/engine/loader/FlutterLoader.java:245]: 512gb -> 512mb

**Comment by dnfield** [comment]
The more I think about this, the more I think we should make this an opt-in parameter. I'm going to change up this PR to do that. We can then look at having some clients set it to whatever level they think is appropriate.

@xster - SGTY?

**Comment by xster** [comment]
Yes I agree

**Comment by dnfield** [comment]
@xster @jason-simmons -  I've refactored this to use half of total memory, or a value specified from the AndroidManifest metadata if available.

This will be more in line with what we do on iOS, and should be less surprising to customers - for example, on many devices today, half of total memory will be 2-6 times as much as the large memory class.

**Comment by chinmaygarde** [comment]
I've reported the flake and re-run the presubmit. This looks good to me and already has approvals. Can we land this?

**Comment by dnfield** [comment]
I think that's fine - if @xster or @jason-simmons have any concerns with the new changes we can revisit.

**Comment by xster** [comment]
This seems like a big change. You're sure we don't want to start with just an opt-in?

**Comment by dnfield** [comment]
I think it's more confusing to have this behavior on ios but not on android.

**Comment by xster** [comment]
LG

**Pull request merged**
